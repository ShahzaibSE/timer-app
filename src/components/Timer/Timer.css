/* Sets the containers height and width */
.base-timer {
    position: relative;
    height: 300px;
    width: 300px;
}
  
/* Removes SVG styling that would hide the time label */
.base-timer__circle {
    fill: none;
    stroke: none;
}
  
/* The SVG path that displays the timer's progress */
.base-timer__path-elapsed {
    stroke-width: 7px;
    stroke: grey;
}

.base-timer__label {
    position: absolute;
    
    /* Size should match the parent container */
    width: 300px;
    height: 300px;
    
    /* Keep the label aligned to the top */
    top: 0;
    
    /* Create a flexible box that centers content vertically and horizontally */
    display: flex;
    align-items: center;
    justify-content: center;
  
    /* Sort of an arbitrary number; adjust to your liking */
    font-size: 48px;
  }

  .base-timer__path-remaining {
    stroke-width: 7px;
    stroke-linecap: round;
    transform: rotate(90deg);
    transform-origin: center;
    transition: 1s linear all;
    fill-rule: nonzero;
    stroke: currentColor;
  }

  .base-timer__path-remaining.green {
    color: rgb(65, 184, 131);
  }

  .base-timer__path-remaining.orange {
    color: orange;
  }
  
  .base-timer__path-remaining.red {
    color: red;
  }

  .base_timer_circle_internal {
    background: #cb2027ad;
    border-radius: 190px;
  }