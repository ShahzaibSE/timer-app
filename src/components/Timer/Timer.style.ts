export const images = [
  {
    url: require("./../../static/images/start-button-background.jpg"),
    title: 'Breakfast',
    width: '35%',
  },
  {
    url: require('./../../static/images/stop-time-background.jpg'),
    title: 'Burgers',
    width: '35%',
  },
  {
    url: require('./../../static/images/reset-button-background.jpg'),
    title: 'Camera',
    width: '35%',
  },
];

