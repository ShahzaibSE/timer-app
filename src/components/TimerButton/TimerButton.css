.timer_button_contianer {
    display: inline;
}