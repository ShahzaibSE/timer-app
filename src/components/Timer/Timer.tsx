import React, {useState, FC} from 'react'
import Grid from "@material-ui/core/Grid"
// Animation utils.
import {AnimationWrapper} from "react-hover-animation"
// Assets.
import "./Timer.css"
import {images} from "./../TimerButton/TimerButton.style"
// Peer component.
import TimerButton from "./../TimerButton/TimerButton"

 // Ring CSS config.,
 const COLOR_CODES = {
    info: {
      color: "green"
    }
};

const FULL_DASH_ARRAY = 283;
// =============== Timer Component ============= //
const Timer: FC = () => {
    // let [timeLeft, setTimeLeft] = useState(60)
    let [isOn, setIsOn] = useState(false)
    let [seconds, setSeconds] = useState(0.0)
    let [intervalID, setIntervalID] = useState(0)
    //
    // CSS colors for ring.
    let remainingPathColor = COLOR_CODES.info.color;
    //
    // Format time.
    const formatTime = (time:any) => {
        // The largest round integer less than or equal to the result of time divided being by 60.
        const minutes = Math.floor(time / 60);
        
        // Seconds are the remainder of the time divided by 60 (modulus operator)
        // setSeconds(time % 60);
        let seconds = time % 60;
        let seconds_formated:any = seconds
        
        // If the value of seconds is less than 10, then display seconds with a leading zero
        if (seconds_formated < 10) {
            seconds_formated = `0${seconds}`;
        }

        // The output in MM:SS format
        return `${minutes}:${seconds_formated}`;
    }
    //
    // Circle decreaser.
    const setCircleDashArray = () => {
        const circleDasharray = `${(
                    seconds * FULL_DASH_ARRAY
                    ).toFixed(0)} 283`;
        //
        let base_timer_path_remaining_element: any|null = document.getElementById("base-timer-path-remaining");
        base_timer_path_remaining_element.setAttribute("strokeDasharray", circleDasharray);
      }
    // setCircleDashArray()  
    //
    // Time features.
    const startTimer = () => {
        console.log("Start timer")
        if (isOn === true) {
            console.log("Can't start timer twice.")
            return
        }
        setIsOn(true)
        let startinterval:any  = setInterval(()=>{
            console.log("Timer running.")
            if (seconds === 0 ){
                setSeconds(++seconds)
            }
        },1000)
        setIntervalID(startinterval)
        setIsOn(true)
    }
    //
    let stopTimer = () => {
        if(isOn === true){
            setIsOn(false)
            clearInterval(intervalID)
            setSeconds(0.0)
            console.log("Stop timer")
        }
    }
    //
    let resetTimer = () => {
        stopTimer()
        setSeconds(0)
    }

    const renderTimeButton = (image:any) => {
        if (image.title === "Reset") {
            let timer_button: any = <Grid item sm={4} md={4} lg={4} key={image.url}> 
                                        <TimerButton key={image.url} buttonTitle={image.title} 
                                        buttonImage={image.url} buttonAction={resetTimer} 
                                        buttonWidth={image.width}/>
                                    </Grid>
            return timer_button
        }else if(image.title === "Stop"){
            let timer_button: any = <Grid item sm={4} md={4} lg={4} key={image.url}> 
                                        <TimerButton key={image.url} buttonTitle={image.title} 
                                        buttonImage={image.url} buttonAction={stopTimer} 
                                        buttonWidth={image.width}/>
                                    </Grid>
            return timer_button
        } else {
            let timer_button: any = <Grid item sm={4} md={4} lg={4} key={image.url}> 
                                        <TimerButton key={image.url} buttonTitle={image.title} 
                                        buttonImage={image.url} buttonAction={startTimer} 
                                        buttonWidth={image.width}/>
                                    </Grid>
            return timer_button
        }
    }
    return (
        <div>
           <Grid container>
            <Grid item sm={12} md={12} lg={12}>    
                <Grid container alignItems="center"
                    justify="center" direction="column" spacing={0} style={{ paddingTop: '12vh' }}>   

                    <Grid item sm={12} md={12} lg={12}>     
                    <div className="base-timer">
                        <svg className="base-timer__svg base_timer_circle_internal" viewBox="0 0 100 100" 
                                xmlns="http://www.w3.org/2000/svg">
                        <g className="base-timer__circle">
                                <circle className="base-timer__path-elapsed" cx="50" cy="50" r="45"></circle>
                                <path
                                    id="base-timer-path-remaining"
                                    strokeDasharray="283"
                                    className={`base-timer__path-remaining ${remainingPathColor}`}
                                    d="
                                    M 50, 50
                                    m -45, 0
                                    a 45,45 0 1,0 90,0
                                    a 45,45 0 1,0 -90,0
                                    "
                                ></path>
                            </g>
                            </svg>
                                <span id="base-timer-label" className="base-timer__label" style={{fontWeight:"bold"}}>
                                        <AnimationWrapper config={{
                                            transform:{initial:'scale(1)',onHover:'scale(1.5)'},
                                            opacity: {initial:'1',onHover:'1'}
                                            }}>
                                        {formatTime(seconds)}
                                    </AnimationWrapper>
                                </span>
                        </div>
                        </Grid>
                    </Grid>
                </Grid>
                <Grid item sm={12} md={12} lg={12}>
                  <div>  
                   <Grid container alignItems="center" justify="center" style={{paddingTop:'7vh'}}>
                         
                        {images.map((image)=>(
                            renderTimeButton(image)
                            // <Grid item sm={4} md={4} lg={4} key={image.url}> 
                            //     <TimerButton key={image.url} buttonTitle={image.title} buttonImage={image.url} buttonAction={()=>{console.log("Button clicked")}} 
                            //     buttonWidth={image.width}/>
                            // </Grid>    
                            ))}
                        
                    </Grid> 
                   </div> 
                 </Grid>
            </Grid>    
        </div>
    )
}

export default Timer
