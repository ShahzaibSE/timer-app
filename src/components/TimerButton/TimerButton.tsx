import React, {FC} from 'react'
import PropTypes from "prop-types"
import ButtonBase from '@material-ui/core/ButtonBase';
import Typography from '@material-ui/core/Typography';
// Animation utils.
import {AnimationWrapper} from "react-hover-animation"
// Assets.
import {timerButtonStyles, images} from "./TimerButton.style"

type TimerButtonProps = {
    buttonTitle: string
    buttonAction: (e:React.MouseEvent<HTMLButtonElement>) => void
    buttonImage: string,
    buttonWidth: string
}

const TimerButton = ({buttonTitle, buttonImage, buttonWidth, buttonAction}: TimerButtonProps) => {
    const classes = timerButtonStyles()
    console.log(images[0].url)
    return (
        <AnimationWrapper config={{transform:{initial:'scale(1)',onHover:'scale(1.2)'},
                                    opacity: {initial:'1',onHover:'1'}}}>
        <div className={classes.root}>
                <ButtonBase
                focusRipple
                key={buttonTitle}
                className={classes.image}
                focusVisibleClassName={classes.focusVisible}
                style={{
                    width: buttonWidth
                }}
                >   
                <span
                    className={classes.imageSrc}
                    style={{
                    backgroundImage: `url('${buttonImage}')`,
                    }}
                />
                <span className={classes.imageBackdrop} />
                <span className={classes.imageButton}>
                    <Typography
                    component="span"
                    variant="subtitle1"
                    color="inherit"
                    className={classes.imageTitle}
                    >
                    {buttonTitle}
                    <span className={classes.imageMarked} />
                    </Typography>
                </span>
                </ButtonBase>
        </div>
        </AnimationWrapper>
    )
}

TimerButton.propTypes = {
    buttonAction: PropTypes.func.isRequired,
    buttonTitle: PropTypes.string.isRequired,
    buttonImage: PropTypes.string.isRequired,
}

export default TimerButton
