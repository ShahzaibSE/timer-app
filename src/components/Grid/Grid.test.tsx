import React from "react"
import { shallow, mount } from "enzyme"
import Timer from "../Timer/Timer"
import TimerButton from "../TimerButton/TimerButton"
// Assets.
import {images} from "./../Timer/Timer.style"

describe("Grid", () => {
    let container: any
    let container_btn: any

    beforeEach(()=>
        container = mount(<Timer/>
    ))

    it('should render a div', () => {
        expect(container.find('div').length).toBeGreaterThanOrEqual(1)
    })

    it('checking presence of Grid component', () => {
        expect(container.containsMatchingElement(<Timer/>)).toEqual(true)
    })

    it('should render timer button', ()=>{
        expect(container.find("TimerButton").length).toBeGreaterThanOrEqual(1)
    })

    // Testing Start, Stop and Reset functionalities.
    // *** //
    // ******** Start Timer ******** //
    // **** //
    // --- Checking If start timer button exist - UI Snapshot testing --- //
    it("checking if start button exist", ()=>{
        container.update()
        expect(container.find("Start")).toMatchSnapshot()
    })

    it("invoke start timer when start timer button is clicked", ()=>{
        const start_timer_mock_fn = jest.fn() // Mock Start timer function.
        const startTimer = mount(<TimerButton 
                                    buttonTitle={images[0].title} 
                                    buttonImage={images[0].url} 
                                    buttonWidth={images[0].width}
                                    buttonAction={start_timer_mock_fn} />)
        const startBtn = startTimer.find("button").first()
        startBtn.update()
        expect(start_timer_mock_fn).toHaveBeenCalledTimes(0)
        startBtn.simulate("click")
        expect(start_timer_mock_fn).toHaveBeenCalledTimes(1)
    })

    // *** //
    // ******** Stop Timer ******** //
    // **** //
    // --- Checking If stop timer button exist - UI Snapshot testing --- //
    it("checking if stop button exist", ()=>{
        container.update()
        expect(container.find("Stop")).toMatchSnapshot()
    })

    it("invokes stopTimer when the stop button is clicked", ()=>{
        const stop_timer_mock_fn = jest.fn() // Mock Start timer function.
        const stopTimer = mount(<TimerButton 
                                    buttonTitle={images[1].title} 
                                    buttonImage={images[1].url} 
                                    buttonWidth={images[1].width}
                                    buttonAction={stop_timer_mock_fn} />)
        const stopBtn = stopTimer.find("button").first()
        stopBtn.update()
        expect(stop_timer_mock_fn).toHaveBeenCalledTimes(0)
        stopBtn.simulate("click")
        expect(stop_timer_mock_fn).toHaveBeenCalledTimes(1)
    })

    // *** //
    // ******** Reset Timer ******** //
    // **** //
    // --- Checking if reset timer button exist - UI Snapshot testing --- //
    it("checking if reset button exist", ()=>{
        container.update()
        expect(container.find("Reset")).toMatchSnapshot()
    })

    it("invokes resetTimer when the reset button is clicked", ()=>{
        const reset_timer_mock_fn = jest.fn() // Mock Start timer function.
        const resetTimer = mount(<TimerButton 
                                    buttonTitle={images[2].title} 
                                    buttonImage={images[2].url} 
                                    buttonWidth={images[2].width}
                                    buttonAction={reset_timer_mock_fn} />)
        const resetBtn = resetTimer.find("button").first()
        resetBtn.update()
        expect(reset_timer_mock_fn).toHaveBeenCalledTimes(0)
        resetBtn.simulate("click")
        expect(reset_timer_mock_fn).toHaveBeenCalledTimes(1)
    })

    // Checking if 
    it("checking if isOn state is changing to true", ()=>{
        container.update()
        expect(container.find("Start")).toMatchSnapshot()
    })

})