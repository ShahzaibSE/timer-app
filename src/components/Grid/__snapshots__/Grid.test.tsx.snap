// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid checking if isOn state is changing to true 1`] = `null`;

exports[`Grid checking if isOn state is changing to true 2`] = `null`;

exports[`Grid checking if isOn state is changing to true 3`] = `null`;

exports[`Grid checking if reset button exist 1`] = `null`;

exports[`Grid checking if start button exist 1`] = `null`;

exports[`Grid checking if stop button exist 1`] = `null`;
